themes/default/content/blog/repairing-state-with-pulumi-refresh/index.md

```diff
@@ -0,0 +1,98 @@
+# Max length is 160 characters.
+meta_desc: Walk through of how Pulumi's state can diverge from provider reality, and how `pulumi refresh` can correct Pulumi's state.
+
+meta_image: meta.png
+
+authors:
+    - ian-wahbe
+
+tags:
+    - features
+---
+
+Under the hood, Pulumi is a desired state engine. This means that you tell Pulumi what you want, Pulumi knows what already exists, and it makes targeted changes to match the state of the world with your desired state. This works great as long as Pulumi understands the state of the world, which it nearly always does. We will discuss how `pulumi refresh` can be used to bring Pulumi's state back inline with external state.
+
+<!--more-->
+
+## The Problem
+
+There are scenarios where Pulumi’s understanding of the world differs from what actually exists. For Pulumi to work correctly, it needs to reconcile its known state with the actual state. Before we explore how Pulumi solves this problem, we should explore how the Pulumi state gets out of sync with the actual state of the world. There are two main ways for Pulumi to get out of sync: out of bound changes and terminating Pulumi mid-operation. I will explain each in turn.
+
+### Out of bound changes
+
+Pulumi expects to be the sole controller for the resources it provisions. An out of bound change happens when this principle is violated. This is best illustrated with an example. Imagine creating an S3 bucket with Pulumi:
+
+```typescript
+const bucket = new aws_native.s3.Bucket("my-bucket", {
+    tags: [{
+            key: "pulumi-bucket",
+            value: "not important",
+        }],
+});
+```
+
+Then someone goes into the aws console and deletes the bucket. What happens if you then go change a tag on the bucket:
+
+```typescript
+const bucket = new aws_native.s3.Bucket("my-bucket", {
+    tags: [{
+            key: "pulumi-bucket",
+            value: "don’t delete me",
+        }],
+});
+```
+
+Pulumi has recorded creating a bucket, but it does not know that the bucket was deleted. It will try to update the bucket and fail, since the bucket does not exist anymore.
+
+### Terminating Pulumi Mid-Operation
+
+Pulumi also gets confused if the CLI or Pulumi Provider process is terminated while an update is taking place. The Pulumi CLI checks in its state before and after each operation it performs. Creating the bucket from above goes something like:
+
+a. Started “pulumi up”.
+b. About to create `aws_native:s3:Bucket::my-bucket`.
+c. Created `aws_native:s3:Bucket::my-bucket` with ID `my-bucket-6e3d099`.
+d. Finished “pulumi up”.
+
+If the Pulumi CLI process is terminated between steps (b) and (c), then Pulumi does not know if
+“my-bucket” exists or not. Unlike (1), Pulumi is aware that it does not know if the resource was created. This will be displayed as a pending CREATE operation. Depending on what type of operation was interrupted, you can see pending CREATE, UPDATE, DELETE, READ and IMPORT operations.
```

Comment:
```suggestion
“my-bucket” exists or not. Unlike the first example, Pulumi is aware that it does not know if the resource was created. This will be displayed as a pending CREATE operation. Depending on what type of operation was interrupted, you can see pending CREATE, UPDATE, DELETE, READ and IMPORT operations.
```